core/controllers/editor_test.py

```diff
@@ -1381,6 +1392,64 @@ def test_for_assign_role_for_exploration(self):
         self.assert_cannot_voiceover(exp_id)
         self.logout()
 
+        # Check that COLLABORATOR 4, who cannot edit but can voiceover,
+        # cannot add a new State 6.
+        self.login(self.COLLABORATOR4_EMAIL)
+        csrf_token = self.get_new_csrf_token()
+
+        # Since check edit_activity_always return False, therefore
+        # user will behave as voiceover artist.
+        get_voiceover_swap = self.swap_to_always_return(
+            rights_manager, 'check_can_edit_activity', value=False)
+
+        with get_voiceover_swap:
+            self.assert_cannot_edit(exp_id)
+            self.assert_can_voiceover(exp_id)
+
+            self.put_json(
+                '%s/%s' % (feconf.EXPLORATION_DATA_PREFIX, exp_id),
+                {
+                    'version': exploration.version,
+                    'commit_message': 'Added State 4',
+                    'change_list': [{
+                        'cmd': 'add_state',
+                        'state_name': 'State 4'
+                    }]
+                },
+                csrf_token=csrf_token,
+                expected_status_int=500
+            )
+
+            reader_dict = self.get_json(
```

Comment:
Here and below: you should have a pre-check too, before the action, to demonstrate that this post-check is the same as the pre-check. See the general guidance here:

https://github.com/oppia/oppia/wiki/Backend-tests#guidelines-for-writing-good-tests

Reply:
I added pre-checks. Thanks for the link!